### ISSUE TITLE
[HOLD for payment 2024-03-29] [HOLD for payment 2024-03-26] [$250] [Feature] RNW Image: Add support for http properties in source

### ISSUE THREAD

**Issue by roryabraham** [Expensify/App#12603]
## Problem:

Adding headers to images works in React native, but when this gets converted to Web & Desktop code (via `react-native-web`) it doesn't work - see https://github.com/necolas/react-native-web/issues/1019

## Solution:

Fix this, since we're planning to start sending `encryptedAuthToken` in chat report image requests via header!

cc @kidroca since you've accepted the challenge to take this on.

**Comment by Beamanator** [comment]
Removing `Help Wanted` since @kidroca is gathering some thoughts & plans to get going on this. I'll assign him once he posts in this issue 😅

**Comment by kidroca** [comment]
Hey guys, the linked ticket actually had a PR and it looks like it was almost merged: https://github.com/necolas/react-native-web/pull/1470/files

As you can see it change a bit too much 

There are a few things that I'd like to do differently 
1. Start by adding support for headers - so nothing about supporting POST (yet)
2. Don't change `ImageUriCache` (yet) - the existing PR uses everything in the `ImageSoruce` to create unique key for cache, but this is incorrect IMO - headers of a GET request should not matter for cache (same as browser cache - if you make the same GET request with different headers you get results from cache (when available)) 

Then continue with anything else that has to be implemented: 
1. Add the ability to handle POST requests - that's the easy part
2. Decide whether react-native-web should cache those - this makes a likely unnecessary job hard - as you know POST requests aren't typically cached - they aren't considered idempotent. The easiest thing would be to just not attempt to cache those 

So that's the brief plan

**Comment by puneetlath** [comment]
I added the `NewFeature` label since _I believe_ this is new functionality, not a bug with existing functionality. And I demoted to weekly as a result. @Beamanator If that's incorrect, let me know!

**Comment by Beamanator** [comment]
@kidroca thanks for the outline of your plan, sounds great to me 👍 I also don't understand _why_ we would need to cache post requests (or if it's a good idea) so it sounds like that's something we don't need to worry about, they can implement that themselves if they want 🤷 

Just to be clear, will you make a PR in the main project **and** in our fork?

**Comment by Beamanator** [comment]
@puneetlath that makes sense to me 👍 It's technically a bug, but we haven't used that functionality in the past so it didn't affect us - so it's kinda 1/2 1/2 :D

**Comment by kidroca** [comment]
> Just to be clear, will you make a PR in the main project and in our fork?

Yes, I'll do that

**Comment by Beamanator** [comment]
@thomas-coldwell & I have been chatting a bit about this issue, @kidroca do you know if this fix would also work if we used `react-native-fast-image` in Web / desktop code? a.k.a. do you think `react-native-web` would treat Image components from `react-native` similar to the Image component from `react-native-fast-image`?

Wondering b/c we think would be nice to use `react-native-fast-image` everywhere in the App (assuming we can get it to work w/out any issues / limitations) so the codebase is a bit more consistent & readable. Does that make sense to you too?

**Comment by kidroca** [comment]
@Beamanator 
This might work if `react-native-fast-image` entry point exports different code by platform
- android -> Glide
- ios -> SDWebImage
- others -> fallback to react-native Image (webpack transforms `react-native` imports to `react-native-web` 

Actually it seems there's a flag about it: https://github.com/DylanVann/react-native-fast-image#fallback-boolean
When `fallback` is `true` `FastImage` would use react-native `Image` , for example: 
```jsx
<FastImage fallback={Platform.os == 'web'} />
```

---

It would be best to have explicit support for web - `react-native-fast-image` might list `-web` as a peer dependency and use it, 
if at some point they need more control over stuff that's encapsulated in `-web` they can create their custom component (and stop using `react-native-web` )

---

I think we can easily test this theory by creating a custom App Image component with multiple exports
- index.native.js -> `<FastImage {...props} />`
- index.js -> `<FastImage fallback {...props} />` 

If that works out ok, we can leave it like that or we can push a few changes for `-fast-image` to do the same behind the scenes

**Comment by Beamanator** [comment]
> I think we can easily test this theory by creating a custom App Image component with multiple exports
>
> - index.native.js -> `<FastImage {...props} />`
> - index.js -> `<FastImage fallback {...props} />`
> If that works out ok, we can leave it like that or we can push a few changes for -fast-image to do the same behind the scenes

Ok nice so if this works, we could potentially make a pretty simple `Image` wrapper component that has the files you mentioned ^, yeah? And maybe as a Part 2, submit a PR to `react-native-fast-image` to get web working (maybe with the same strategy).

I'm just thinking we can get it working ASAP in App, then we can clean up later. I think it's better to implement the fix in App quick instead of waiting for a PR to get merged in `react-native-fast-image` or to make a local fork

**Comment by Beamanator** [comment]
So @kidroca it seems like we are going with this step by step plan:
1. You're making web & desktop cache images by fixing `react-native-web` (this issue)
2. @thomas-coldwell is upgrading `iOS` (& maybe Android) caching via `react-native-fast-image` (https://github.com/Expensify/App/pull/12648)

Once these are implemented, "caching images should work". Then we can work on:
1. Using `react-native-fast-image` everywhere (which means using `fallback` prop for Web / Desktop)
2. Clear out / crush related issues in the image enhancements tracker: https://github.com/Expensify/App/issues/10894

**Comment by thomas-coldwell** [comment]
If you can give me a ping when the RNW side of things is sorted @kidroca I'll go through and ensure support using fast image across the whole app as per @Beamanator comment 🙌

**Comment by kidroca** [comment]
This is my PR, when it's reviewed I'll open the same PR for the `react-native-web` repo 
- https://github.com/Expensify/react-native-web/pull/8

I've stumbled on 2 issues that we'd need to address to make this work in App

#### ImageWithSizeCalculation

This component uses the `getSize` static method that works with only `uri` - no source with headers 
https://github.com/Expensify/App/blob/9a8291ba97c381b651b5344d6fa471c51ae70336/src/components/ImageWithSizeCalculation.js#L70

I've made the `onLoad` event to return the image size like it works in RN so we can use that 

#### CORS issues on localhost

> Access to fetch at 'https://www.expensify.com/?exitTo=%2Fchat-attachments%2F949802747%2Fw_1426cdf2308f4ec256096c4dc216242423efdada.jpg' (redirected from 'http://localhost:8080/chat-attachments/949802747/w_1426cdf2308f4ec256096c4dc216242423efdada.jpg') from origin 'http://localhost:8080' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.

I get this error because when I make a request with headers - perhaps this case is not handled by the local proxy, or there's something else...

**Comment by Beamanator** [comment]
Super pumped about this, sadly I haven't had time to review today (catching up on lots of things while i was OOO & the brain is super foggy from intense jetlag) - I'll definitely get to this tomorrow, sorry for the delay!

**Comment by kidroca** [comment]
Opened a few PRs:
- for Expensify fork of RNW - https://github.com/Expensify/react-native-web/pull/8
- for mainstream RNW repo - https://github.com/necolas/react-native-web/pull/2442
- for App - https://github.com/Expensify/App/pull/13036
    This one showcases Image requests with headers working for Web/Desktop.

**Comment by kidroca** [comment]
> 1. Using `react-native-fast-image` everywhere (which means using `fallback` prop for Web / Desktop)

I've tried that, but because `react-native-fast-image` tries to load some native modules in it's init script it's not possible to import it on Web (or at least I couldn't easily find a way) 
So it seems we can't just use the `fallback` prop
I think we should just use the RNW Image component for web/desktop

I did manage to get Images with headers load on web, so we can move forward with what we've got

**Comment by necolas** [comment]
> Fix this, since we're planning to start sending encryptedAuthToken in chat report image requests via header!

How would you expect to do this if image loading worked as it does natively on the web, i.e., just props on `<img>`?

**Comment by Beamanator** [comment]
@necolas I believe in this case we'd try to go with the approach of using `fetch` to send the image get request, with the auth token in a header, then setting the response as the image's source - similar to what's described [here](https://www.twelve21.io/how-to-access-images-securely-with-oauth-2-0/#:~:text=request%20forgery%20(CSRF).-,3.%20USE%20DATA%20URIS,-Yet%20another%20option). I see there's a few cons though, but they seem unconfirmed

**Comment by Beamanator** [comment]
Making progress on upstream PR: https://github.com/necolas/react-native-web/pull/2442

**Comment by JmillsExpensify** [comment]
Looks like that PR has stalled a bit?

**Comment by kidroca** [comment]
> Looks like that PR has stalled a bit?

Yes, I'm waiting for maintainer feedback/review for the PR on the main repo

**Comment by necolas** [comment]
The PR hasn't "stalled". I'm busy working through dozens of other changes for RN and RNWeb that are going to be part of the 0.19 release and were planned months before that PR was posted.

**Comment by JmillsExpensify** [comment]
@necolas I'm super sorry! I didn't mean for that comment to come across as it unfortunately did. We're super appreciative of what you've done with react-native-web, and my comment came from us being eager to help move things forward. If there is any way we can improve how we collaborate from here on out, we're very open to your thoughts and suggestions. It's admittedly been a balance managing upstream fixes that benefit everyone versus our own fork. Thanks again!

**Comment by Beamanator** [comment]
@kidroca In order to unblock our other image enhancement issues, here's our new plan:
1. Work on getting https://github.com/Expensify/react-native-web/pull/8 merged into our fork
2. Get that update working in our App (which will unblock numerous other issues around image caching)
3. Open a new issue tracking the upstream PR so when it gets merged, we can pull those changes into our fork

**Comment by kidroca** [comment]
Allright @Beamanator 
Since I had to squash my changes to fix conflicts in the original repo, I'll have to do the same for Expensify's fork 

> Can you let me know if the fork PR is ready to be tested in App? 🙏

The logic is already established and we verified it here: https://expensify.slack.com/archives/C01GTK53T8Q/p1669800005882449?thread_ts=1669058284.677789&cid=C01GTK53T8Q
But I'll have to apply the latest changes (the squash) to be in sync with the original repo 

The PR we used as a POC: https://github.com/Expensify/App/pull/13036 is using code that was reverted 
In order to make my work testable in App, I'll need that same code available again, maybe it already is - I'll post back if I have any difficulties otherwise I should be done by tomorrow

**Comment by Beamanator** [comment]
> The PR we used as a POC: https://github.com/Expensify/App/pull/13036 is using code that was reverted
In order to make my work testable in App, I'll need that same code available again, maybe it already is - I'll post back if I have any difficulties otherwise I should be done by tomorrow

@kidroca it looks like you're talking about https://github.com/Expensify/App/pull/12648 which got reverted, right? There's a draft follow-up here: https://github.com/Expensify/App/pull/13304 (which also involves a PR to `react-native-fast-image` here: https://github.com/DylanVann/react-native-fast-image/pull/953)

**Comment by kidroca** [comment]
@Beamanator 
I've updated the PR to our fork with latest code 
I've also cherry picked everything I need as a single commit for the PR for App here: https://github.com/Expensify/App/pull/13036
This way we can test the Image web/desktop logic in App - when the other PR (fast-image) is ready I'll drop the "cherry commit" and use the actual code, but for web/desktop there should be no difference 

My only problem right now is that I can't login - getting 403s - I've submitted my email and IP to be whitelisted

**Comment by Beamanator** [comment]
Thanks @kidroca 👍 

Sadly I'm actually going OOO till Jan 1, maybe @roryabraham could you help test this out before I get back, or maybe @marcaaron ?

**Comment by kidroca** [comment]
Reposting my last summary as it's still valid (https://github.com/Expensify/App/issues/12603#issuecomment-1327517367)

Opened a few PRs:
- for Expensify fork of RNW - https://github.com/Expensify/react-native-web/pull/8 - this is ready for review ✔️ 
- for mainstream RNW repo - https://github.com/necolas/react-native-web/pull/2442 - I don't think this is fully reviewed yet
- for App - https://github.com/Expensify/App/pull/13036 - ATM this is a draft showcasing Image requests with headers working for Web/Desktop. 

I guess someone needs to review the PR for our fork so we can merge it in order to start using it in App
- https://github.com/Expensify/react-native-web/pull/8

**Comment by Beamanator** [comment]
Thanks @kidroca - sorry for the delay, I was OOO for a bit over the holidays.

I just did a review on the Expensify fork - overall it looks great but it would be great to get a few things cleared up (some in GH comments, others in code comments) before we move forward - I'm also asking Marc to re-review & see if we need another set of 👀

**Comment by kidroca** [comment]
I've addressed PR comments and we're close to a final version of the PR

**Comment by kidroca** [comment]
I've verified the latest PR changes are still working OK in App, though I did discover this CORS issue by accident: https://expensify.slack.com/archives/C049HHMV9SM/p1673289214225929

It seems that because an attachment was sent from STAGING it needs to be loaded from the STAGING url (even though I'm on a different ENV)

This + HEADERS qualifies the request as a CORS request and creates an issue (because the backend response does not satisfy CORS rules)

I think it would be best if we can skip the `origin` from the chat attachment URL, this would make the request against the current origin, so instead of saving `https://staging.expensify.com/chat-attachments/...atachment.jpg` we save just `/chat-attachments/...atachment.jpg` and use that as source or at least add the prefix/origin ourselves

**Comment by Beamanator** [comment]
Little updates on this issue:
- [Fork PR](https://github.com/Expensify/react-native-web/pull/11) was merged 🎉 
- [App PR](https://github.com/Expensify/App/pull/13036) was updated and should be ready for review
- [Upsteam PR](https://github.com/necolas/react-native-web/pull/2442) has some activity - @kidroca when you get a chance can you please respond to the latest comment? It looks like the maintainer is asking for more tests

**Comment by kidroca** [comment]
> * [Upsteam PR](https://github.com/necolas/react-native-web/pull/2442) has some activity - @kidroca when you get a chance can you please respond to the latest comment? It looks like the maintainer is asking for more tests

I need to sync the latest changes to the upstream PR, and write some unit tests

**Comment by MelvinBot** [comment]
This issue has not been updated in over ***15*** days. @puneetlath, @Beamanator, @kidroca eroding to ***Monthly*** issue.

P.S. Is everyone reading this sure this is really a near-term priority? Be brave: if you disagree, go ahead and close it out. If someone disagrees, they'll reopen it, and if they don't: one less thing to do!

**Comment by puneetlath** [comment]
@Beamanator I'm going to unassign myself from this since I don't think we need a BZ on this one. Feel free to assign me back if I'm wrong about that.

**Comment by kidroca** [comment]
The [Upstream PR](https://github.com/necolas/react-native-web/pull/2442#pullrequestreview-1282972631) is updated and it's ready for review

Meanwhile the current task is blocked by 
- https://github.com/Expensify/App/issues/14991

Should we should update the PR and issue title with `[HOLD]` ?

**Comment by trjExpensify** [comment]
Yeah, I think we should update the title of this issue if it's blocked: `[Hold - #14991]`

**Comment by johnmlee101** [comment]
Do we believe this is a dupe? https://github.com/Expensify/App/issues/15922 where the loading of attachments on the viewer is blocked by the caching this implements?

**Comment by kidroca** [comment]
> Do we believe this is a dupe? https://github.com/Expensify/App/issues/15922 where the loading of attachments on the viewer is blocked by the caching this implements?

I don't think so 

The spinner probably appears because we render it until the `onLoad` event is fired 

Even when we have image in cache it would still fire this event

_(I'll double check and post back)_

**Comment by kidroca** [comment]
@johnmlee101 I've double checked and we're not a blocker - images are still getting cached for at least 2hrs
An image loads from cache if we go back to it
 <img width="567" alt="Screenshot 2023-03-17 at 17 00 59" src="https://user-images.githubusercontent.com/12156624/226046702-a9d54d53-c9ad-4195-876a-5257ea94aec2.png">

**Comment by MelvinBot** [comment]
This issue has not been updated in over ***15*** days. @Beamanator, @kidroca eroding to ***Monthly*** issue.

P.S. Is everyone reading this sure this is really a near-term priority? Be brave: if you disagree, go ahead and close it out. If someone disagrees, they'll reopen it, and if they don't: one less thing to do!

**Comment by Beamanator** [comment]
This one's still on hold but maybe we can keep it `Monthly` since we're holding on a bigger project

**Comment by trjExpensify** [comment]
Can we link to the right place for the hold on this issue, please? The issue title references #14991 which has merged, so it's confusing as to what the dependency really is. Thanks!

**Comment by trjExpensify** [comment]
Commented [here](https://github.com/Expensify/App/issues/10894#issuecomment-1503257531) on that.

**Comment by trjExpensify** [comment]
Reopening this as it's still important to do, just held.

**Comment by Beamanator** [comment]
Yikes i wonder if I have to comment for melvin to reset the overdue time

**Comment by trjExpensify** [comment]
Sneaky Melv, this isn't done.

**Comment by kidroca** [comment]
@trjExpensify there is a PR open (for App), that implements the changes needed to cache images 

- I need to sync it with the main branch as the PR is very old 
- Confirm that feature (images with http headers) work now that the other issue is resolved
- Find someone to review and merge the PR

**Comment by trjExpensify** [comment]
Sounds great! We can ask a C+ to jump on it when it's ready, and @Beamanator is already assigned from our side internally. 

>I need to sync it with the main branch as the PR is very old

My thoughts are with you my friend! 😅 🙏

**Comment by Beamanator** [comment]
Nice, ya looks like you're talking about this PR (https://github.com/Expensify/App/pull/13036), right @kidroca ?

I am happy to help test & review with @aimane-chnaif once it's ready - just let us know 🙏

**Comment by kidroca** [comment]
Everything seems to be working OK now!
The PR is now ready for review
- https://github.com/Expensify/App/pull/13036#pullrequestreview-1714534306

**Comment by trjExpensify** [comment]
Wahooo nice! Excited for it!

**Comment by trjExpensify** [comment]
Nice catch, done!

**Comment by kidroca** [comment]
A new problem came up - we've switched to the upstream RNW library
- ✔️ CORS issue is resolved
- ❌ we're no longer use our `react-native-web` fork which had support for Image with HTTP headers
- 🟡 the feature is planned for v0.20 in the upstream `react-native-web` repo
    - https://github.com/necolas/react-native-web/pull/2442#issuecomment-1502287124

A couple of days earlier when I tested this we still used our fork, but recently we've merged this PR which updates RNW and switches to the upstream repo
- https://github.com/Expensify/App/pull/24482

Our `package.json` when 
https://github.com/Expensify/App/blob/12966ae5c5954fd7bb056a5ac39dc17a768c8064/package.json#L162

`package.json` before:
https://github.com/Expensify/App/blob/07727226efabeaef9069e9d9423572ec9f0352e7/package.json#L56

cc @Beamanator

**Comment by getusha** [comment]
@kidroca can't we apply the the changes in https://github.com/necolas/react-native-web/pull/2442 as a patch?

**Comment by kidroca** [comment]
> @kidroca can't we apply the the changes in [necolas/react-native-web#2442](https://github.com/necolas/react-native-web/pull/2442) as a patch?

That's an idea

Can you remind me how do we make sure my patch isn't wiping any other patch we might have for `react-native-web`, or how do we make sure someone else in the future doesn't wipe my own patch, by applying another patch at some point? 

Do we have the process documented?

**Comment by getusha** [comment]
@kidroca i don't think it's documented anywhere, but, we can create a separate patch for each PR, it doesn't have to be all in one file. there are around 3 patch files for rnw currently.

**Comment by kidroca** [comment]
Hey @trjExpensify,

We've encountered an issue detailed in https://github.com/Expensify/App/issues/12603#issuecomment-1804450624

As @getusha suggested in [this comment](https://github.com/Expensify/App/issues/12603#issuecomment-1807683188), we can use `patch-package` to apply our modifications to `react-native-web`. This approach aligns with our existing strategy, given that we currently maintain 4 patches for `react-native-web`.

I believe applying as patch is a practical solution since we no longer have a fork of react-native-we However, I would appreciate official confirmation to ensure alignment with our project's guidelines and to avoid any unforeseen complications.

Thank you for your guidance on this matter.

**Comment by trjExpensify** [comment]
That sounds okay to me! IIRC, we just need to also track when we can remove the patch in the future. CC: @roryabraham

**Comment by Beamanator** [comment]
I like the patch idea as well - @trjExpensify I think we can just create a GH issue to follow up with https://github.com/necolas/react-native-web/pull/2442 & once it's merged, we can see if any other patches are fixed & if we can remove them by bumping our rnw version

**Comment by kidroca** [comment]
Ok, the PR is updated with a patch for react-native-web 
- https://github.com/Expensify/App/pull/13036
- ✔️ Ready for Review

**Comment by Beamanator** [comment]
> Cool, got a specific process to follow for that? Who gets assigned etc.

Hmm no, I don't think we have an official process for this - we're kinda "figuring it out as we go" - I have this issue where I've been tracking a few upstream PRs to forks - https://github.com/Expensify/App/issues/14382

But it has had no progress in about a year lol, I'm not sure what the next steps would be, since we can't get any info / response from the maintainer of `react-native-fast-image` - but at least we do get some responses from the maintainer of `react-native-web`!

**Comment by trjExpensify** [comment]
Okay cool, so would you create one of those for RNW and co-assign with @kidroca or something?

**Comment by Beamanator** [comment]
Created tracker here: https://github.com/Expensify/App/issues/32544

I linked PRs that created the existing 4 patches so we can track them too, though I couldn't find "upstream PRs" for them, which seems non-ideal - i think ideally we'd discuss the patches w/ the maintainer & eventually try to get the patches merged into the main codebase or have the maintainer accept some other solution for fixes - either way, I think we can continue this convo in the new linked issue, as it doesn't relate to this specific issue

**Comment by Beamanator** [comment]
Seems this got reverted 🤔  https://github.com/Expensify/App/pull/32700

**Comment by Beamanator** [comment]
Seems we're now holding on this, before we can re-implement the changes: https://github.com/Expensify/App/issues/32703

**Comment by blimpich** [comment]
https://github.com/Expensify/App/issues/32703 was fixed, should be good to go forward on this now

**Comment by kidroca** [comment]
> #32703 was fixed, should be good to go forward on this now

Just to be clear - I just need to post the same PR that was reverted due to #32703, anything else?

**Comment by blimpich** [comment]
> Just to be clear - I just need to post the same PR that was reverted due to #32703, anything else?

From my understanding, yes, that should be all that is needed. As per your comment [here](https://github.com/Expensify/App/issues/32593#issuecomment-1845364895), we updated the `/receipts/{path}` for CORS requests to be the same as the way `/chat-attachments/{path}` was updated. Sorry it took so long.

**Comment by trjExpensify** [comment]
@aimane-chnaif adding you as the C+ to this issue btw, now that you're on reviewing the PR.

**Comment by trjExpensify** [comment]
Going to remove the payment hold on this, as I believe we need to address what happened causing the revert.

**Comment by trjExpensify** [comment]
👋 @kidroca do you have an ETA for the next steps here? Thanks!

**Comment by kidroca** [comment]
> 👋 @kidroca do you have an ETA for the next steps here? Thanks!

Currently there's no way to recreate the issue as it only happen when the code is deployed to `staging` 
Once the issue in thread shared by @aimane-chnaif  https://expensify.slack.com/archives/C01GTK53T8Q/p1707417010918089?thread_ts=1706812735.378739&cid=C01GTK53T8Q is resolved, we can deploy to a PR testing domain which would allow us to debug and fix the underlying issue (and also we'll be able to catch such issues early)

**Comment by trjExpensify** [comment]
Perfect, appreciate that. The thread wasn't on my radar earlier today 👍

**Comment by trjExpensify** [comment]
Copying the [next step](https://expensify.slack.com/archives/C01GTK53T8Q/p1707824961768889?thread_ts=1706812735.378739&cid=C01GTK53T8Q) from the thread for posterity: 

alexbeaman 
So I wonder if the next step would be - @kidroca if you could make another PR, then we test on the adhoc build to hopefully make sure everything is working perfectly? :pray:

**Comment by kidroca** [comment]
Opened a new PR
- #36560 

Let's see if we can recreate the issue with receipts images once the pr testing domain is available, following the steps here: https://github.com/Expensify/App/pull/35636#issuecomment-1922941224

**Comment by Beamanator** [comment]
Thanks @kidroca ! I triggered an adhoc build for that PR, it should be ready in an hour ish 👍

**Comment by trjExpensify** [comment]
Looking pretty good here, no revert! What's next? We move to https://github.com/Expensify/App/issues/9402? :excited:

**Comment by Beamanator** [comment]
@trjExpensify do you mind helping figure out payments here so we can close this out? 😅 🙏  I'm not sure who was the previous BugZero person assigned here, but looks like we got none now :O

**Comment by trjExpensify** [comment]
@aimane-chnaif offer sent! Let me know when you accept.

**Comment by trjExpensify** [comment]
Paid!
